aspects/schema.go

```diff
@@ -0,0 +1,467 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package aspects
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"regexp"
+
+	"github.com/snapcore/snapd/strutil"
+)
+
+func ParseSchema(raw []byte) (*CustomSchema, error) {
+	var schemaDef map[string]json.RawMessage
+	err := json.Unmarshal(raw, &schemaDef)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse top level schema: top level must be a map")
+	}
+
+	schema := &CustomSchema{}
+	if val, ok := schemaDef["types"]; ok {
+		var userTypes map[string]json.RawMessage
+		if err := json.Unmarshal(val, &userTypes); err != nil {
+			return nil, fmt.Errorf(`cannot parse user-defined types at top level (must be a map): %w`, err)
+		}
+
+		schema.userTypes = make(map[string]Schema, len(userTypes))
+		for userTypeName, typeDef := range userTypes {
+			userTypeSchema, err := schema.Parse(typeDef)
+			if err != nil {
+				return nil, fmt.Errorf(`cannot parse user-defined type %q: %w`, userTypeName, err)
+			}
+
+			schema.userTypes[userTypeName] = userTypeSchema
+		}
+	}
+
+	schema.topLevel, err = schema.Parse(raw)
+	if err != nil {
+		return nil, err
+	}
+
+	return schema, nil
+}
+
+type CustomSchema struct {
+	userTypes map[string]Schema
+	topLevel  Schema
+}
+
+func (s *CustomSchema) Validate(raw []byte) error {
+	return s.topLevel.Validate(raw)
+}
+
+func (s *CustomSchema) Parse(raw json.RawMessage) (Schema, error) {
+	var typ string
+	var schemaDef map[string]json.RawMessage
+	var hasConstraints bool
+	if err := json.Unmarshal(raw, &schemaDef); err != nil {
+		var typeErr *json.UnmarshalTypeError
+		if !errors.As(err, &typeErr) {
+			return nil, err
+		}
+
+		if err := json.Unmarshal(raw, &typ); err != nil {
+			return nil, err
+		}
+	} else {
+		// schema definition is a map, we might have constraints to process
+		hasConstraints = true
+
+		rawType, ok := schemaDef["type"]
+		if !ok {
+			typ = "map"
+		} else {
+			if err := json.Unmarshal(rawType, &typ); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	var schema Schema
+	switch typ {
+	case "map":
+		schema = &mapSchema{
+			topSchema: s,
+		}
+	case "int":
+		schema = &intSchema{}
+	case "string":
+		schema = &stringSchema{}
+	case "number":
+		return nil, nil
+	case "bool":
+		return nil, nil
+	case "array":
+		return nil, nil
+	default:
+		if typ[0] != '$' {
+			return nil, fmt.Errorf("cannot parse type %q: unknown", typ)
+		}
+		userType, ok := s.userTypes[typ[1:]]
+		if !ok {
+			return nil, fmt.Errorf(`cannot find referenced user-defined type %q`, typ)
+		}
+		schema = userType
+	}
+
+	// only parse the schema if it's a map definition w/ constraints
+	if hasConstraints {
+		if err := schema.Parse(raw); err != nil {
+			return nil, err
+		}
+	}
+
+	return schema, nil
+}
+
+type stringSchema struct {
```

Comment:
we want all the schema struct to carry some kind of name/path composed while parsing using the keys/types/positions under which the schema appear, we want to use those names/paths then consistently in the Validate error messages